### PR TITLE
MAKE-610: gridfs using new godriver

### PR DESCRIPTION
**Pull request by tychoish**

### CODE REVIEW

**Comment by tychoish** [comment]
This is still in progress, pending:
- we need to vendor the new go driver at the new location
- add this to the current testing infrastructure

File: util.go [hunk 1]
```diff
@@ -73,6 +73,20 @@ func walkLocalTree(ctx context.Context, prefix string) ([]string, error) {
 	return out, nil
 }
 
+func walkLocalTreeMap(ctx context.Context, prefix string) (map[string]struct{}, error) {
```
Comment (julianedwards) [inline util.go]: Is this function used?

File: gridfs.go [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+		return nil
+	} else if err != nil {
+		return errors.Wrap(err, "problem finding file")
+	}
+
+	document := struct {
+		ID interface{} `bson:"_id"`
+	}{}
+
+	for cursor.Next(ctx) {
```
Comment (julianedwards) [inline gridfs.go]: Why is this written as a duplicate of the RemoveMany function?

File: gridfs.go [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+	iter, err := b.List(ctx, remote)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	keys := []string{}
+
+	for iter.Next(ctx) {
+		item := iter.Item()
+		name := filepath.Join(local, item.Name()[len(remote)+1:])
```
Comment (julianedwards) [inline gridfs.go]: I would use the filepath.Rel function -> `filepath.Rel(remote, item.Name())`

File: gridfs.go [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+	writer, err := grid.OpenDownloadStreamByName(name)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem opening stream")
+	}
+
+	return writer, nil
+}
+
+func (b *gridfsBucket) Put(ctx context.Context, name string, input io.Reader) error {
+	if b.opts.DryRun {
```
Comment (julianedwards) [inline gridfs.go]: I would move this if clause to line 122, after attempting to resolve the bucket

File: gridfs.go [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+	if ok {
+		_ = gfs.SetReadDeadline(dl)
+		_ = gfs.SetWriteDeadline(dl)
+	}
+
+	return gfs, nil
+}
+
+func (b *gridfsBucket) Writer(ctx context.Context, name string) (io.WriteCloser, error) {
+	if b.opts.DryRun {
```
Comment (julianedwards) [inline gridfs.go]: I would move this to after the if clause that resolves the bucket

File: gridfs.go [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+	return writer, nil
+}
+
+func (b *gridfsBucket) Reader(ctx context.Context, name string) (io.ReadCloser, error) {
+	grid, err := b.bucket(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem resolving bucket")
+	}
+
+	writer, err := grid.OpenDownloadStreamByName(name)
```
Comment (julianedwards) [inline gridfs.go]: should this variable be called `reader` instead of `writer`?

File: gridfs.go (hunk 2 above)
Comment (tychoish) [reply gridfs.go]: Remove handles missing keys differently.

File: gridfs.go (hunk 3 above)
Comment (tychoish) [reply gridfs.go]: I think this would lead to weird outcomes on windows

File: gridfs.go (hunk 2 above)
Comment (julianedwards) [reply gridfs.go]: so filename is not unique here?

File: gridfs.go [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+		return errors.Wrap(err, "problem iterating gridfs metadata")
+	}
+	if err = cursor.Close(ctx); err != nil {
+		return errors.Wrap(err, "problem closing ")
+	}
+
+	return nil
+}
+
+func (b *gridfsBucket) RemoveMany(ctx context.Context, keys ...string) error {
```
Comment (julianedwards) [inline gridfs.go]: Need dry run here

File: gridfs.go (hunk 2 above)
Comment (tychoish) [reply gridfs.go]: It might not be.

Comment (julianedwards) [reply gridfs.go]: got it

**Pull request merged**